RecoEcal/EgammaCoreTools/src/EcalClusterLazyTools.cc

```diff
@@ -22,76 +22,56 @@
 #include "RecoLocalCalo/EcalRecAlgos/interface/EcalSeverityLevelAlgoRcd.h"
 #include "RecoLocalCalo/EcalRecAlgos/interface/EcalSeverityLevelAlgo.h"
 
-EcalClusterLazyToolsBase::EcalClusterLazyToolsBase( const edm::Event &ev, const edm::EventSetup &es, edm::EDGetTokenT<EcalRecHitCollection> token1, edm::EDGetTokenT<EcalRecHitCollection> token2) {
-
-  ebRHToken_ = token1;
-  eeRHToken_ = token2;
- 
-  getGeometry( es , false );
-  getTopology( es );
-  getEBRecHits( ev );
-  getEERecHits( ev );
+namespace {
+
+  template<class T, class R>
+  auto getFromEventSetup(edm::EventSetup const& eventSetup) {
+    edm::ESHandle<T> handle;
+    eventSetup.get<R>().get(handle);
+    return handle.product();
+  }
+
+  auto getRecHits(edm::Event const& event, edm::EDGetTokenT<EcalRecHitCollection> const& token) {
+    edm::Handle< EcalRecHitCollection > recHitsHandle;
+    event.getByToken( token, recHitsHandle );
+    return recHitsHandle.product();
+  }
+
+};
+
+EcalClusterLazyToolsBase::EcalClusterLazyToolsBase( const edm::Event &ev, const edm::EventSetup &es, edm::EDGetTokenT<EcalRecHitCollection> token1, edm::EDGetTokenT<EcalRecHitCollection> token2)
+  : geometry_(getFromEventSetup<CaloGeometry, CaloGeometryRecord>(es))
+  , topology_(getFromEventSetup<CaloTopology, CaloTopologyRecord>(es))
+  , ebRecHits_(getRecHits(ev, token1))
+  , eeRecHits_(getRecHits(ev, token2))
```

Comment:
Let me also note that instead of a `EcalRecHitCollection`-specific function one could use
https://github.com/cms-sw/cmssw/blob/2421a1f398e0a8f5f5a2419fb1679240652c66a9/FWCore/Framework/interface/Event.h#L705-L711
with `eeRecHits_(&edm::get(ev, token2))` etc.

Reply:
Nice! So how come I have not seen this function before, that is really a better interface than the geByToken thing! Thanks for pointing that out!

Reply:
Also the EventSetup version seems to exist already:

https://github.com/fwyzard/cmssw/blob/master/FWCore/Framework/interface/EventSetup.h#L167